### PR TITLE
[JENKINS-56063] (Expansion of variables in refspec in case of honor refspec on initial clone)

### PR DESCRIPTION
**Pull request by rishabhBudhouliya**
## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by rishabhBudhouliya** [comment]
I haven't added tests yet, would add them if this is the right approach to fix this issue.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 1]
```diff
@@ -148,7 +149,20 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
             // in a single job definition.
             RemoteConfig rc = scm.getRepositories().get(0);
             List<RefSpec> refspecs = rc.getFetchRefSpecs();
-            cmd.refspecs(refspecs);
+            List<RefSpec> expandedrefSpecs = new ArrayList<>();
```
Comment (proski) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: The name capitalization should be `expandedRefSpecs` for camelCase.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 2]
```diff
@@ -148,7 +149,20 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
...
+            for (RefSpec ref:refspecs) {
+                if(ref.toString().contains("$")){
+                    EnvVars env = build.getEnvironment(listener);
+                    expandedrefSpecs.add(new RefSpec(env.expand(ref.toString())));
+                    check = true;
+                }
+            }
+            if(!check) {
+                cmd.refspecs(refspecs);
+            } else{
```
Comment (proski) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: I'm not sure if git-pugin has coding style, but the spacing for braces is inconsistent.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 3]
```diff
@@ -148,7 +149,20 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
             // in a single job definition.
             RemoteConfig rc = scm.getRepositories().get(0);
             List<RefSpec> refspecs = rc.getFetchRefSpecs();
-            cmd.refspecs(refspecs);
+            List<RefSpec> expandedrefSpecs = new ArrayList<>();
+            Boolean check = false;
```
Comment (proski) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: `check` is a poor name. I don't see what is being checked here.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 4]
```diff
@@ -148,7 +149,20 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
...
+            for (RefSpec ref:refspecs) {
+                if(ref.toString().contains("$")){
```
Comment (proski) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: If one `ref` doesn't contain `$`, it won't be added to `expandedrefSpecs`. However, it another `ref` contains `$`, `expandedrefSpecs` would still be used. Is that intentional?

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 5]
```diff
@@ -148,7 +149,20 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
...
+                    EnvVars env = build.getEnvironment(listener);
```
Comment (proski) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: Shouldn't this be done outside the loop?

Comment (rishabhBudhouliya) [reply src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: I have changed the implementation and shifted this outside.

**Comment by rishabhBudhouliya** [comment]
@proski Thanks for reviewing my PR, after going through your suggestions I have changed the implementation which is more coherent to how variables are expanded in GitSCM.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java (hunk 4 above)
Comment (rishabhBudhouliya) [reply src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: No it wasn't intentional, this logic wouldn't cover all the cases, thanks for pointing that out!

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 6]
```diff
@@ -148,7 +149,12 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
             // in a single job definition.
             RemoteConfig rc = scm.getRepositories().get(0);
             List<RefSpec> refspecs = rc.getFetchRefSpecs();
-            cmd.refspecs(refspecs);
+            List<RefSpec> expandedRefSpecs = new ArrayList<>();
+            EnvVars env = build.getEnvironment(listener);
+            for (RefSpec ref:refspecs) {
```
Comment (proski) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: Please use spaces around `:`, they are used in most places in the code, and they make the code more readable.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 7]
```diff
@@ -148,7 +149,12 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
...
+                expandedRefSpecs.add(new RefSpec(env.expand(ref.toString())));
+            }
+                cmd.refspecs(expandedRefSpecs);
```
Comment (proski) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: The indent is wrong here.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java (hunk 6 above)
Comment (rishabhBudhouliya) [reply src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: Thanks for the advice, I have corrected it.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 8]
```diff
@@ -148,7 +149,12 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
             // in a single job definition.
             RemoteConfig rc = scm.getRepositories().get(0);
             List<RefSpec> refspecs = rc.getFetchRefSpecs();
-            cmd.refspecs(refspecs);
+            List<RefSpec> expandedRefSpecs = new ArrayList<>();
+            EnvVars env = build.getEnvironment(listener);
+            for (RefSpec ref : refspecs) {
```
Comment (MarkEWaite) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: Needs automated tests to confirm:
* Operating system environment variables (like PATH) are expanded as expected (PATH will generate a nonsense refspec, but it has the benefit of being defined on all supported platforms)
* Jenkins variables (like [GIT_BRANCH](https://plugins.jenkins.io/git#branch-variables)) are expanded as expected in a Freestyle project
* Jenkins token macros (like [GIT_REVISION](https://plugins.jenkins.io/token-macro)) are expanded as expected in a Freestyle project

Comment (rishabhBudhouliya) [reply src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: I have added a test file which is running successful tests for "BUILD_ID", "BUILD_NO" and "GIT_COMMIT". 
I have two issues here: 

- I can't make a freestyle build with ${PATH} added in the refspec, it fails the build with a "_can't file the commitFile in the working directory_". Although I have tested this **interactively** where you can clearly see that the refspec is expanded but the build fails, I still would want your opinion on this.
Screenshot 2020-02-07 at 9 19 29 PM

- By Jenkins token macros do you mean running a build with parameters? Is that how tokens are used? If yes then how does one set a parameter for a job? Should I put the variable in the EnvVars store?

Comment (rishabhBudhouliya) [reply src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: I've seen the Token Macro API but I don't think that servers our purpose here 
`assertThat(TokenMacro.expandAll(build2, listener, "${GIT_REVISION}"), is(sha1String));`

Comment (rishabhBudhouliya) [reply src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: After doing some RCA I found out that adding ${PATH} to refspec would make it illegal, hence the job will fail before doing anything else.

**Comment by MarkEWaite** [comment]
Changes from this pull request were also included in #841 but were then reverted from #841.  This pull request still needs to be reviewed and merged before the change will be available to users.

**Comment by jetersen** [comment]
What is blocking this issue? We recently updated to v4 and found jobs that no longer worked.

**Comment by MarkEWaite** [comment]
Needs review and merge by me or @rsandell or @fcojfernandez .  Interactive testing to confirm that it is behaves correctly with:
* operating system environment variables
* variables set as parameters by job definition
* environment variables defined by env-inject plugin

**Comment by jetersen** [comment]
After digging into the problem our problem was actually related to #772 #898

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java (hunk 8 above)
Comment (fcojfernandez) [reply src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java:154]: > I can't make a freestyle build with ${PATH} added in the refspec, it fails the build with a "can't file the commitFile in the working directory"

How did you tried? Just adding `${PATH}` to `permuteRefSpecVariable`?

I'd also like to know how this PR behaves with job parameters (https://wiki.jenkins.io/display/JENKINS/Define+Parameters)

**Comment by MarkEWaite** [comment]
@rishabhBudhouliya I restarted my investigation of the issue this is describing.  I have pushed a rebased version of this pull request as [rebased-JENKINS-56063-Fix](https://github.com/MarkEWaite/git-plugin/pull/new/rebased-JENKINS-56063-Fix).  When master is merged to this branch or when this is rebased on the master branch, test failures are visible, just as they are visible in the CI job.

If you have time to investigate further, I can leave this pull request open.

If you don't have time to investigate further, I'll close this pull request and keep the branch on my private fork for future work.

I've also added test cases to my [docker-lfs repository](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/jobs/Bugs-Individual/jobs/Bugs-50-000-to-59-999/jobs/JENKINS-56063-refspec-env-reference-not-expanded/config.xml), [docker-lfs with jgit](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/jobs/Bugs-Individual/jobs/Bugs-50-000-to-59-999/jobs/JENKINS-56063-refspec-env-reference-not-expanded-jgit/config.xml) and my [jenkins-bugs repository](https://github.com/MarkEWaite/jenkins-bugs/tree/JENKINS-56063-refspec-env-reference-not-expanded) so that the bug will be regularly assessed during automated testing and exploratory.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Test failures need to be fixed.  My interactive testing confirmed that (as described in the original report), command line git behaves OK because the shell expands the environment variables even though the git plugin does not expand them.  It also confirmed that JGit fails because it does not expand the variables and does not invoke a shell.

**Comment by MarkEWaite** [comment]
Closing until someone has time to fix the test failures.

**Pull request closed without merging**